### PR TITLE
- added program to create a subvolume

### PR DESCRIPTION
**Pull request by aschnell**

### CODE REVIEW

File: client/mksubvolume.cc [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+
+bool set_nocow = false;
+bool verbose = false;
+
+
+void
+do_tmp_mount(const libmnt_fs* fs, const char* tmp_mountpoint, const string& subvol_option)
+{
+    if (verbose)
+	cout << "do-tmp-mount" << endl;
```
Comment (shundhammer) [inline client/mksubvolume.cc]: How about using __FUNCTION__ for this, preferably in a macro?
e.g.

#define TRACE() if (verbose) cout << __FUNCTION__ << endl

File: client/mksubvolume.cc [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+
+    libmnt_fs* x = mnt_copy_fs(NULL, fs);
```
Comment (shundhammer) [inline client/mksubvolume.cc]: Poor variable naming. I'd prefer something like 'fs', 'filesystem', 'filesys'. I had to look it up several times what again this 'x' is.

File: client/mksubvolume.cc [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+    if (!x)
+	throw runtime_error("mnt_copy_fs failed");
+
+    struct libmnt_context* cxt = mnt_new_context();
+    mnt_context_set_fs(cxt, x);
+    mnt_context_set_target(cxt, tmp_mountpoint);
+    if (!subvol_option.empty())
+	mnt_context_set_options(cxt, ("subvol=" + subvol_option).c_str());
+    else
+	mnt_context_set_options(cxt, "subvolid=5");
```
Comment (shundhammer) [inline client/mksubvolume.cc]: really literal 5 here?

File: client/mksubvolume.cc [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+void
+do_create_subvolume(const string& tmp_mountpoint, const string& subvolume_name)
+{
+    if (verbose)
+	cout << "do-create-subvolume" << endl;
+
+    string parent = tmp_mountpoint + "/" + dirname(subvolume_name);
+
+    if (mkdir(parent.c_str(), 0777) != 0 && errno != EEXIST)
+	throw runtime_error_with_errno("mkdir failed", errno);
```
Comment (shundhammer) [inline client/mksubvolume.cc]: not ST_THROW() since this was introduced here in other places?

File: client/mksubvolume.cc [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+
+    int fd = open(parent.c_str(), O_RDONLY);
+    if (fd < 0)
+	throw runtime_error_with_errno("open failed", errno);
```
Comment (shundhammer) [inline client/mksubvolume.cc]: again: not ST_THROW() ?

File: client/mksubvolume.cc [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+
+void
+do_tmp_umount(const libmnt_fs* fs, const char* tmp_mountpoint)
+{
+    if (verbose)
+	cout << "do-tmp-umount" << endl;
+
+    system("/sbin/udevadm settle --timeout 20");
+
+    libmnt_fs* x = mnt_copy_fs(NULL, fs);
```
Comment (shundhammer) [inline client/mksubvolume.cc]: same complaint as above: better use 'fs' etc. instead of 'x'

File: client/mksubvolume.cc (hunk 3 above)
Comment (aschnell) [inline client/mksubvolume.cc]: Yes, that is the initially top-level subvolume with btrfs.

File: client/mksubvolume.cc (hunk 1 above)
Comment (aschnell) [inline client/mksubvolume.cc]: Sure, but on the other hand I don't like defines.

File: client/mksubvolume.cc (hunk 4 above)
Comment (aschnell) [inline client/mksubvolume.cc]: The program does not setup logging functions so ST_THROW does not bring benefits.

File: client/mksubvolume.cc (hunk 2 above)
Comment (aschnell) [inline client/mksubvolume.cc]: Unfortunately the parameter is already called fs. Having e.g. fs and filesystem is not better.

**Comment by shundhammer** [comment]
LGTM

**Pull request merged**